Fix callback concurrency issues; ensure more methods will resolve callbacks; do a best-case handling for userCompletedAction; update readme

pragunvohra opened the pull request:
Had some trouble with the Cordova SDK when testing with Android, so here are some fixes.

1. No more class-level callbackContext; individual callbackContexts are
passed between methods to ensure we don’t clobber the callbackContext
with two concurrent calls to execute()

2. logout, registerView, and userCompletedAction were missing callback
resolutions, so I added those. userCompletedAction’s logic is not
completely correct as it will always (early) resolve to “success” but
Branch.java in the Branch SDK does not expose a method that takes in a
listener for userCompletedAction that I saw.

3. Fixed CreditHistoryListener to reject a promise if there was a JSON
exception

4. Added properly evaluated callback to setIdentity

5. Updated Readme.md to match code for installation

aaustin commented:
So awesome @pragunvohra !! We'll have a chance to review this weekend, and hopefully merge this in over the next few days. In the meantime, can you help yourself to a Branch sweatshirt? https://branch.io/branch-referral-program-claim-reward/

pragunvohra commented:
:+1: great. Let me know if you have any questions. I sent a more detailed email to support@branchmetrics.io with the issues here along with some other ones I found that I did not fix, but should be addressed.

Anu2g commented:
Hey all, I work with Pragun. To be clear, we'll need you guys (Branch) to fix these 3 items in order for this plugin to be production-ready. Do you know when you guys would be able to fix that by? We'd like to be able to use this in production ASAP. Thanks!

aaustin commented:
Hey guys - thanks for the heads up. We unfortunately are relying on a third party contractor for the issues so I can't guarantee that we'll get this done today. However, I really appreciate the detailed spec, since it'll help them resolve the issues.

I'll relay the message and description right now. I really appreciate your contribution.

harshabonthu commented:
@pragunvohra Regarding issue#1 i.e dependency not being added, Plugin will be adding that automatically (https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/blob/master/plugin.xml#L73). Ive tested with a new cordova project -> http://recordit.co/sLguz9jx3H

renesansz commented:
Hey @pragunvohra ,
Thanks for the heads up, currently I'm trying to figure out issue #2. It's impossible to return an instance of the branch universal object from Java to Javascript since both of them handle objects differently.
That's the reason why I only instantiate the branch universal link in Java and return a JSON response to Javascript, then once successfully instantiated I attached the branch universal object methods.
I'm trying to figure out how you can create multiple instance of the object since plugin's Java class is only instantiated once Cordova has launched, at the moment you cannot create multiple instance of the branch universal object.

pragunvohra commented:
@renesansz, one thought is to have the creation method return an id for a branch object, while keeping a map of <id, BUO> in BranchSDK.java; then, when calling methods on the branch object, you take in an ID and use the appropriate branch object by pulling it from the <id, BUO> map.

What you can do in your plugin's js file is take the branch universal object id in as a return value from Java and append some methods on a new object that would execute the remainder of the cordova plugin functions on the branch universal object (or somehow wraps the other branch universal object methods) before passing that object to the success function handler.

Does that make sense?

pragunvohra commented:
@harshabonthu, glad that it worked in your test case. Not sure why it didn't work when we added it. All I did was run:

cordova plugin add https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK.git --variable BRANCH_KEY=<your-branch-key> --variable BRANCH_TEST_KEY=<your-branch-key> --variable URI_SCHEME=<your-app-uri-scheme-without-colon-and-slashes>

but I'm not familiar enough with the cordova processes to know if I had to do anything else, thus I just added the SDK reference to the grade file for #1 manually.

renesansz commented:
@pragunvohra oh, I see, so I need to have an array to hold out the generated universal objects, then I can do the api transactions by passing the id as reference for the object instance.

Yes, I've already done appending the universal branch methods before the object is being returned.

Thanks for the suggestion, everything is clear to me now.

pragunvohra commented:
Exactly :+1:

renesansz commented:
@pragunvohra , just have created a [PR](https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/pull/76) to address the issues you've mentioned.

Let me know if it solves your issues with the plugin.

aaustin commented:
I've got the thumbs up to merge this to master.
